File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    int answer = 0;
    public int solution(int k, int[][] dungeons) {
        boolean[] visited = new boolean[dungeons.length];
        explore(k, dungeons, visited, 0);
        return answer;
    }

    private void explore(int k, int[][] dungeons, boolean[] visited, int i) {
        answer = Math.max(answer, i);
        for(int j = 0; j < dungeons.length; j++) {
            if(!visited[j] && k >= dungeons[j][0]) {
                visited[j] = true;
                explore(k - dungeons[j][1], dungeons, visited, i + 1);
                visited[j] = false;
            }
        }
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    explore_comb = list(permutations(dungeons))
    arr_cnt = []    

    for comb in explore_comb:
        fatigue = k
        cnt = 0

        for dungeon in comb:
            if explore_dungeon(fatigue, dungeon[0], dungeon[1]) != -1:
                cnt += 1
                fatigue -= dungeon[1]

        arr_cnt.append(cnt)

        if max(arr_cnt) == len(dungeons):
            break

    return max(arr_cnt)

def explore_dungeon(remain_fatigue : int, min_fatigue : int, use_fatigue : int):
    if remain_fatigue < min_fatigue:
        return -1
    
    return remain_fatigue - use_fatigue